### PR TITLE
Bugfix/restriction values are now validated. 

### PR DESCRIPTION
**Pull request by YannickWeber**
**Motivation**
It was possible to enter negative values into all inner values of the 'restrictions'-tag in the configuration.

**Changes**
Those values are now validated, when an invalid value is given, the default value is used. Also a warning is printed.

### CODE REVIEW

File: src/main/java/com/hivemq/configuration/reader/RestrictionConfigurator.java [hunk 1]
```diff
@@ -31,12 +36,56 @@ public RestrictionConfigurator(@NotNull final RestrictionsConfigurationService r
...
+        if (maxConnections < MAX_CONNECTIONS_MINIMUM) {
+            log.warn("The configured max-connections ({}) must be greater than 0. The value was set to {} (unlimited) instead.", maxConnections, MAX_CONNECTIONS_DEFAULT);
+            return MAX_CONNECTIONS_DEFAULT;
+        }
+        return maxConnections;
+    }
+
+    private int validateMaxClientIdLength(final int maxClientIdLength) {
+        if (maxClientIdLength < MAX_CLIENT_ID_LENGTH_MINIMUM || maxClientIdLength > MAX_CLIENT_ID_LENGTH_MAXIMUM) {
+            log.warn("The configured max-clientid-length ({}) must be in the range {} - {}. The value was set to {} instead.", MAX_CLIENT_ID_LENGTH_MINIMUM, MAX_CLIENT_ID_LENGTH_MAXIMUM, maxClientIdLength, MAX_CLIENT_ID_LENGTH_DEFAULT);
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/configuration/reader/RestrictionConfigurator.java]:
```suggestion
            log.warn("The configured max-clientid-length ({}) must be in the range {} - {}. The value was set to {} instead.", maxClientIdLength, MAX_CLIENT_ID_LENGTH_MINIMUM, MAX_CLIENT_ID_LENGTH_MAXIMUM, MAX_CLIENT_ID_LENGTH_DEFAULT);
```

File: src/main/java/com/hivemq/configuration/reader/RestrictionConfigurator.java [hunk 2]
```diff
@@ -31,12 +36,56 @@ public RestrictionConfigurator(@NotNull final RestrictionsConfigurationService r
...
+            return MAX_CLIENT_ID_LENGTH_DEFAULT;
+        }
+        return maxClientIdLength;
+    }
+
+    private int validateMaxTopicLength(final int maxTopicLength) {
+        if (maxTopicLength < MAX_TOPIC_LENGTH_MINIMUM || maxTopicLength > MAX_TOPIC_LENGTH_MAXIMUM) {
+            log.warn("The configured max-topic-length ({}) must be in the range {} - {}. The value was set to {} instead.", MAX_TOPIC_LENGTH_MINIMUM, MAX_TOPIC_LENGTH_MINIMUM, maxTopicLength, MAX_TOPIC_LENGTH_DEFAULT);
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/configuration/reader/RestrictionConfigurator.java]:
```suggestion
            log.warn("The configured max-topic-length ({}) must be in the range {} - {}. The value was set to {} instead.", maxTopicLength, MAX_TOPIC_LENGTH_MINIMUM, MAX_TOPIC_LENGTH_MINIMUM, MAX_TOPIC_LENGTH_DEFAULT);
```

File: src/main/java/com/hivemq/configuration/reader/RestrictionConfigurator.java [hunk 3]
```diff
@@ -31,12 +36,56 @@ public RestrictionConfigurator(@NotNull final RestrictionsConfigurationService r
...
+        if (noConnectIdleTimeout < NO_CONNECT_IDLE_TIMEOUT_MINIMUM) {
+            log.warn("The configured no-connect-idle-timeout ({}) must be greater than 0. The value was set to {} instead.", noConnectIdleTimeout, NO_CONNECT_IDLE_TIMEOUT_DEFAULT);
+            return NO_CONNECT_IDLE_TIMEOUT_DEFAULT;
+        }
+        return noConnectIdleTimeout;
     }
 
+    private long validateIncomingLimit(final long incomingLimit) {
+        if (incomingLimit < INCOMING_BANDWIDTH_THROTTLING_MINIMUM) {
+            log.warn("The configured incoming-bandwidth-throttling ({} bytes/second) must not be negative. The value was set to {} bytes/second instead.", incomingLimit, INCOMING_BANDWIDTH_THROTTLING_DEFAULT);
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/configuration/reader/RestrictionConfigurator.java]: Why does this message contain units and the others do not?

Comment (SgtSilvio) [inline src/main/java/com/hivemq/configuration/reader/RestrictionConfigurator.java]:
```suggestion
            log.warn("The configured incoming-bandwidth-throttling ({} bytes/second) must not be negative. The value was set to unlimited instead.", incomingLimit);
```

File: src/test/java/com/hivemq/configuration/reader/RestrictionConfiguratorTest.java [hunk 4]
```diff
@@ -17,13 +17,18 @@
 package com.hivemq.configuration.reader;
 
 import com.google.common.io.Files;
+import io.netty.util.internal.UnstableApi;
```
Comment (SgtSilvio) [inline src/test/java/com/hivemq/configuration/reader/RestrictionConfiguratorTest.java]: Why was this import added?

File: src/main/java/com/hivemq/configuration/reader/RestrictionConfigurator.java (hunk 2 above)
Comment (YannickWeber) [reply src/main/java/com/hivemq/configuration/reader/RestrictionConfigurator.java]: You are right, I made a mistake here. But your suggestion is also incorrect. Fixed it anyway.

File: src/main/java/com/hivemq/configuration/reader/RestrictionConfigurator.java (hunk 3 above)
Comment (YannickWeber) [reply src/main/java/com/hivemq/configuration/reader/RestrictionConfigurator.java]: no-connect-idle-timout can have 'ms' as a unit.

File: src/main/java/com/hivemq/configuration/reader/RestrictionConfigurator.java [hunk 5]
```diff
@@ -31,12 +36,65 @@ public RestrictionConfigurator(@NotNull final RestrictionsConfigurationService r
...
+        restrictionsConfigurationService.setIncomingLimit(validateIncomingLimit(restrictionsEntity.getIncomingBandwidthThrottling()));
+        restrictionsConfigurationService.setMaxTopicLength(validateMaxTopicLength(restrictionsEntity.getMaxTopicLength()));
+    }
+
+    private long validateMaxConnections(final long maxConnections) {
+        if (maxConnections == UNLIMITED_CONNECTIONS) {
+            return maxConnections;
+        }
+        if (maxConnections < MAX_CONNECTIONS_MINIMUM) {
+            log.warn("The configured max-connections ({}) must be greater than 0. The value was set to {} (unlimited) instead.", maxConnections, MAX_CONNECTIONS_DEFAULT);
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/configuration/reader/RestrictionConfigurator.java]:
```suggestion
            log.warn("The configured max-connections ({}) must be greater than {}. The value was set to {} (unlimited) instead.", maxConnections, MAX_CONNECTIONS_MINIMUM, MAX_CONNECTIONS_DEFAULT);
```

File: src/main/java/com/hivemq/configuration/reader/RestrictionConfigurator.java [hunk 6]
```diff
@@ -31,12 +36,65 @@ public RestrictionConfigurator(@NotNull final RestrictionsConfigurationService r
...
+                    maxTopicLength, MAX_TOPIC_LENGTH_MINIMUM, MAX_TOPIC_LENGTH_MAXIMUM, MAX_TOPIC_LENGTH_DEFAULT);
+            return MAX_TOPIC_LENGTH_DEFAULT;
+        }
+        return maxTopicLength;
+    }
+
+    private long validateNoConnectIdleTimeout(final long noConnectIdleTimeout) {
+        if (noConnectIdleTimeout < NO_CONNECT_IDLE_TIMEOUT_MINIMUM) {
+            log.warn(
+                    "The configured no-connect-idle-timeout ({}ms) must be greater than 0ms. The value was set to {}ms instead.",
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/configuration/reader/RestrictionConfigurator.java]:
```suggestion
                    "The configured no-connect-idle-timeout ({} ms) must be greater than 0ms. The value was set to {}ms instead.",
```

File: src/main/java/com/hivemq/configuration/reader/RestrictionConfigurator.java [hunk 7]
```diff
@@ -31,12 +36,65 @@ public RestrictionConfigurator(@NotNull final RestrictionsConfigurationService r
...
+                    noConnectIdleTimeout, NO_CONNECT_IDLE_TIMEOUT_DEFAULT);
+            return NO_CONNECT_IDLE_TIMEOUT_DEFAULT;
+        }
+        return noConnectIdleTimeout;
     }
 
+    private long validateIncomingLimit(final long incomingLimit) {
+        if (incomingLimit < INCOMING_BANDWIDTH_THROTTLING_MINIMUM) {
+            log.warn(
+                    "The configured incoming-bandwidth-throttling ({} bytes/second) must not be negative. The value was set to {} (unlimited) instead.",
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/configuration/reader/RestrictionConfigurator.java]:
```suggestion
                    "The configured incoming-bandwidth-throttling ({} bytes/second) must not be negative. It was set to the default (unlimited) instead.",
```

File: src/main/java/com/hivemq/configuration/reader/RestrictionConfigurator.java [hunk 8]
```diff
@@ -31,12 +36,65 @@ public RestrictionConfigurator(@NotNull final RestrictionsConfigurationService r
...
+                    incomingLimit, INCOMING_BANDWIDTH_THROTTLING_DEFAULT);
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/configuration/reader/RestrictionConfigurator.java]:
```suggestion
                    incomingLimit);
```

File: src/main/java/com/hivemq/configuration/reader/RestrictionConfigurator.java [hunk 9]
```diff
@@ -31,12 +36,65 @@ public RestrictionConfigurator(@NotNull final RestrictionsConfigurationService r
...
+            log.warn("The configured max-connections ({}) must be greater than 0. The value was set to {} (unlimited) instead.", maxConnections, MAX_CONNECTIONS_DEFAULT);
+            return MAX_CONNECTIONS_DEFAULT;
+        }
+        return maxConnections;
+    }
+
+    private int validateMaxClientIdLength(final int maxClientIdLength) {
+        if (maxClientIdLength < MAX_CLIENT_ID_LENGTH_MINIMUM || maxClientIdLength > MAX_CLIENT_ID_LENGTH_MAXIMUM) {
+            log.warn(
+                    "The configured max-clientid-length ({}) must be in the range {} - {}. The value was set to {} instead.",
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/configuration/reader/RestrictionConfigurator.java]:
```suggestion
                    "The configured max-clientid-length ({}) must be in the range {} - {}. It was set to the default ({}) instead.",
```

File: src/main/java/com/hivemq/configuration/reader/RestrictionConfigurator.java [hunk 10]
```diff
@@ -31,12 +36,65 @@ public RestrictionConfigurator(@NotNull final RestrictionsConfigurationService r
...
+                    MAX_CLIENT_ID_LENGTH_DEFAULT);
+            return MAX_CLIENT_ID_LENGTH_DEFAULT;
+        }
+        return maxClientIdLength;
+    }
+
+    private int validateMaxTopicLength(final int maxTopicLength) {
+        if (maxTopicLength < MAX_TOPIC_LENGTH_MINIMUM || maxTopicLength > MAX_TOPIC_LENGTH_MAXIMUM) {
+            log.warn(
+                    "The configured max-topic-length ({}) must be in the range {} - {}. The value was set to {} instead.",
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/configuration/reader/RestrictionConfigurator.java]:
```suggestion
                    "The configured max-topic-length ({}) must be in the range {} - {}. It was set to the default ({}) instead.",
```

File: src/main/java/com/hivemq/configuration/reader/RestrictionConfigurator.java (hunk 6 above)
Comment (SgtSilvio) [inline src/main/java/com/hivemq/configuration/reader/RestrictionConfigurator.java]:
```suggestion
                    "The configured no-connect-idle-timeout ({}ms) must be at least {}ms. It was set to the default ({}ms) instead.",
```

File: src/main/java/com/hivemq/configuration/reader/RestrictionConfigurator.java [hunk 11]
```diff
@@ -31,12 +36,65 @@ public RestrictionConfigurator(@NotNull final RestrictionsConfigurationService r
...
+            return MAX_TOPIC_LENGTH_DEFAULT;
+        }
+        return maxTopicLength;
+    }
+
+    private long validateNoConnectIdleTimeout(final long noConnectIdleTimeout) {
+        if (noConnectIdleTimeout < NO_CONNECT_IDLE_TIMEOUT_MINIMUM) {
+            log.warn(
+                    "The configured no-connect-idle-timeout ({}ms) must be greater than 0ms. The value was set to {}ms instead.",
+                    noConnectIdleTimeout, NO_CONNECT_IDLE_TIMEOUT_DEFAULT);
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/configuration/reader/RestrictionConfigurator.java]:
```suggestion
                    noConnectIdleTimeout, NO_CONNECT_IDLE_TIMEOUT_DEFAULT, NO_CONNECT_IDLE_TIMEOUT_MINIMUM);
```

File: src/main/java/com/hivemq/configuration/reader/RestrictionConfigurator.java [hunk 12]
```diff
@@ -31,12 +36,69 @@ public RestrictionConfigurator(@NotNull final RestrictionsConfigurationService r
...
+                    NO_CONNECT_IDLE_TIMEOUT_DEFAULT);
+            return NO_CONNECT_IDLE_TIMEOUT_DEFAULT;
+        }
+        return noConnectIdleTimeout;
     }
 
+    private long validateIncomingLimit(final long incomingLimit) {
+        if (incomingLimit < INCOMING_BANDWIDTH_THROTTLING_MINIMUM) {
+            log.warn(
+                    "The configured incoming-bandwidth-throttling ({} bytes/second) must not be negative. The default value (unlimited) is used instead.",
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/configuration/reader/RestrictionConfigurator.java]:
```suggestion
                    "The configured incoming-bandwidth-throttling ({} bytes/second) must be at least {}. The default value (unlimited) is used instead.",
```

File: src/main/java/com/hivemq/configuration/reader/RestrictionConfigurator.java [hunk 13]
```diff
@@ -31,12 +36,69 @@ public RestrictionConfigurator(@NotNull final RestrictionsConfigurationService r
...
+                    incomingLimit);
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/configuration/reader/RestrictionConfigurator.java]:
```suggestion
                    incomingLimit, INCOMING_BANDWIDTH_THROTTLING_MINIMUM);
```

File: src/main/java/com/hivemq/configuration/reader/RestrictionConfigurator.java [hunk 14]
```diff
@@ -31,12 +36,70 @@ public RestrictionConfigurator(@NotNull final RestrictionsConfigurationService r
...
+                    "The configured no-connect-idle-timeout ({}ms) must be at least {}ms. The default value ({}ms) is used instead.",
+                    noConnectIdleTimeout,
+                    NO_CONNECT_IDLE_TIMEOUT_MINIMUM,
+                    NO_CONNECT_IDLE_TIMEOUT_DEFAULT);
+            return NO_CONNECT_IDLE_TIMEOUT_DEFAULT;
+        }
+        return noConnectIdleTimeout;
     }
 
+    private long validateIncomingLimit(final long incomingLimit) {
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/configuration/reader/RestrictionConfigurator.java]: should be `validateIncomingBandwidthThrottling`
